### PR TITLE
rego file parser

### LINKED ISSUE #1435: Support for Rego
[Rego](https://www.openpolicyagent.org/docs/latest/#rego) is a high level declarative language built for expressing policy over hierarchical data structures. Used in [Open Policy Agent](https://www.openpolicyagent.org/).

`cson` for Atom in the OPA project -
https://github.com/open-policy-agent/opa/blob/master/misc/syntax/atom/language-rego/grammars/rego.cson

### PR DESCRIPTION
**Pull request by deltamualpha**
* Add lexer for [Rego](https://www.openpolicyagent.org/docs/latest/policy-language/#what-is-rego), which is a datalog-like language used to write policy files for authentication/authorization.
* Reference taken from the official [Rego.tmLanguage](https://github.com/open-policy-agent/vscode-opa/blob/master/syntaxes/Rego.tmLanguage) as well as the prolog lexer (very similar language). It's a pretty simple implementation!
* This will close #1435.

### CODE REVIEW

File: lib/rouge/lexers/rego.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+            state :basic do
+              rule %r/\s+/, Text
+              rule %r/#.*/, Comment::Single
+      
+              rule %r/[\[\](){}|.,;!]/, Punctuation
+      
+              rule %r/"[^"]*"/, Str::Double
+      
+              rule %r/-?\d+\.\d+([eE][+-]?\d+)?/, Num::Float
```
Comment (pyrmont) [inline lib/rouge/lexers/rego.rb]: Is a number followed by a period (and nothing else, e.g. `13.`) treated as a number in Rego or as a number and punctuation?

File: lib/rouge/lexers/rego.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+              rule %r/-?\d+([eE][+-]?\d+)?/, Num
+
+              rule %r/\\u[0-9a-fA-F]{4}/, Num::Hex
+              rule %r/\\["\/bfnrt]/, Str::Delimiter
```
Comment (pyrmont) [inline lib/rouge/lexers/rego.rb]: The token for these should be `Str::Escape`.

```suggestion
              rule %r/\\["\/bfnrt]/, Str::Escape
```

File: lib/rouge/lexers/rego.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            state :atoms do
+              rule %r/(true|false|null)/, Keyword::Constant
+              rule %r/[[:word:]]*/, Str::Symbol
+              rule %r/'[^']*'/, Str::Symbol
+            end
+      
+            state :operators do
+              rule %r/(=|!=|>|<|>=|<=|\+|-|\*|%|\/|\||&|:=)/, Operator
+              rule %r/(default|not|package|import|as|with|else|some)/, Operator
+              rule %r/[#&*+-.\/:<=>?@^~]+/, Operator
```
Comment (pyrmont) [inline lib/rouge/lexers/rego.rb]: A number of the operators here seem duplicative of operators in other rules. I think I've removed all the duplicates but please double check.

```suggestion
              rule %r/[\/:?@^~]+/, Operator
```

File: lib/rouge/lexers/rego.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            end
+      
+            state :atoms do
+              rule %r/(true|false|null)/, Keyword::Constant
+              rule %r/[[:word:]]*/, Str::Symbol
+              rule %r/'[^']*'/, Str::Symbol
+            end
+      
+            state :operators do
+              rule %r/(=|!=|>|<|>=|<=|\+|-|\*|%|\/|\||&|:=)/, Operator
```
Comment (pyrmont) [inline lib/rouge/lexers/rego.rb]: These match from left to right so you need a slight reordering:

```suggestion
              rule %r/(=|!=|>=|<=|>|<|\+|-|\*|%|\/|\||&|:=)/, Operator
```

File: spec/visual/samples/rego [hunk 5]
```diff
@@ -0,0 +1,27 @@
+package httpapi.authz
```
Comment (pyrmont) [inline spec/visual/samples/rego]: Could you add some additional examples to this file to make sure the lexing is being done correctly? At the moment there are no numbers, no single-quoted strings, etc.

File: lib/rouge/lexers/rego.rb (hunk 1 above)
Comment (deltamualpha) [reply lib/rouge/lexers/rego.rb]: it'd be a parse error, actually, if used as a bare word. I should see if I can lex that as an error condition...

Comment (pyrmont) [reply lib/rouge/lexers/rego.rb]: No worries. It's a non-goal of Rouge to check for syntax correctness so you don't need to worry too much about ensuring syntactical mistakes are lexed as errors—just wanted to check this wasn't an overlooked corner case :)

File: spec/visual/samples/rego (hunk 5 above)
Comment (deltamualpha) [reply spec/visual/samples/rego]: Added numbers; single-quote strings are actually invalid in the language, so there's no valid lex for them. I think that covers all of the facets of the syntax; it's pretty simple!

**Comment by deltamualpha** [comment]
Committed the rest of the changes. Do I need to smush this down to a single commit?

File: spec/visual/samples/rego (hunk 5 above)
Comment (pyrmont) [reply spec/visual/samples/rego]: I might have used the wrong term here. In the `atoms` state, there's a rule matching single-quoted strings that generates `Str::Symbol` tokens. This is what I had in mind.

**Comment by pyrmont** [comment]
@deltamualpha No need to squash! I do that when I merge and it's more helpful for code review here to have all the commits. Thanks for checking, though! :D

File: spec/visual/samples/rego (hunk 5 above)
Comment (deltamualpha) [reply spec/visual/samples/rego:1]: Ah, I'd mentally glossed over that. I've just removed that rule, since it's not a valid lex to begin with.

**Comment by pyrmont** [comment]
@deltamualpha Thanks for the submission :) This will be part of the next version of Rouge. That's scheduled for release on Tuesday 14 April 🎉

**Pull request merged**